Matrix.C

```diff
@@ -1262,5 +1263,104 @@ const
 #endif
 }
 
+Matrix outerProduct(const Vector &v, const Vector &w)
+{
+    /*
+     * There are two cases of concern:
+     *
+     * 1) w is not distributed
+     *
+     * 2) w is distributed
+     *
+     */
+    int result_num_rows = v.dim();
+    int result_num_cols;
+    bool is_distributed = v.distributed();
+    Vector gathered_w;
+
+    /*
+     * Gather all of the entries in w on each process into a Vector stored
+     * redundantly (i.e., not distributed) on each process. If the Vector w
+     * is not distributed, this step is trivial.
+     */
+    if (!w.distributed())
+    {
+        result_num_cols = w.dim();
+        gathered_w = w;
+    }
+    else // w.distributed() is true
+    {
+        // Get the number of columns on each processor and gather these
+        // counts into an array. Use std::vector as an array substitute
+        // because variable-length arrays aren't in the C++ standard,
+        // but the storage for std::vector containers must be contiguous
+        // as defined by the C++ standard.
+        int process_local_num_cols = w.dim();
+        MPI_Comm comm = MPI_COMM_WORLD;
+        MPI_Datatype num_cols_datatype = MPI_INT;
+        int num_procs;
+        MPI_Comm_size(comm, &num_procs);
+        std::vector<int> num_cols_on_proc(num_procs, 0);
+        int num_procs_send_count = 1;
+        int num_procs_recv_count = 1;
+        MPI_Allgather(&process_local_num_cols,
+                      num_procs_send_count,
+                      num_cols_datatype,
+                      num_cols_on_proc.data(),
+                      num_procs_recv_count,
+                      num_cols_datatype,
+                      comm);
+
+        // Compute the displacements for the entries in the gathered
+        // Vector, along with the total number of columns in the result.
+        std::vector<int> gathered_w_displacements(num_procs, 0);
+        result_num_cols = 0;
+        for (int i = 0; i < num_cols_on_proc.size(); i++)
+        {
+            gathered_w_displacements.at(i) = result_num_cols;
+            result_num_cols += num_cols_on_proc.at(i);
+        }
+
+        // Gather the data from each process onto each process -- this
+        // step is an Allgatherv (because difference processes may
+        // have different numbers of entries.
+        std::vector<double> w_data(process_local_num_cols);
```

Comment:
It seems like it would be nice to have a public `const double* Vector::data()` function to eliminate some copies like this. As a bonus, since LAPACK and BLAS are required anyway, we could then use them to do operations like an outer product faster than the for-loop implementation.